Shared/Samples/Download preplanned map area/DownloadPreplannedMapAreaView.Model.swift

```diff
@@ -0,0 +1,311 @@
+                    offlineMap = mmpk.maps.first
+                } else {
+                    // If we have a failure, then keep the online map selected.
+                    selectedMap = oldValue
+                }
+            }
+        }
+        
+        /// Updates the `canRemoveDownloadedMaps` state.
+        private func updateCanRemoveDownloadedMaps() {
+            canRemoveDownloadedMaps = allOfflineMapModels.contains(where: \.downloadDidSucceed)
+        }
+        
+        /// Cancels all current jobs.
+        private func cancelAllJobs() async {
+            await withTaskGroup(of: Void.self) { group in
+                allOfflineMapModels.forEach { model in
+                    if model.isDownloading {
+                        group.addTask {
+                            await model.cancelDownloading()
+                        }
+                    }
+                }
+            }
+        }
+        
+        // Removes all downloaded maps.
+        func removeDownloadedMaps() {
+            // Sets the current map to the online web map.
+            selectedMap = .onlineWebMap
+            
+            // Update state.
+            canRemoveDownloadedMaps = false
+            
+            Task {
+                // Cancel all current download jobs.
+                await cancelAllJobs()
+                
+                // Remove each download.
+                allOfflineMapModels.forEach { $0.removeDownloadedContent() }
+            }
+        }
+    }
+}
+
+extension DownloadPreplannedMapAreaView.Model {
+    /// A type that specifies the currently selected map.
+    enum SelectedMap: Hashable {
+        /// The online version of the map.
+        case onlineWebMap
+        /// One of the offline maps.
+        case offlineMap(OfflineMapModel)
+    }
+}
+
+/// An object that encapsulates state about an offline map.
+class OfflineMapModel: ObservableObject, Identifiable {
+    /// The preplanned map area.
+    let preplannedMapArea: PreplannedMapArea
+    /// The task to use to take the area offline.
+    let offlineMapTask: OfflineMapTask
+    /// The directory where the mmpk will be stored.
+    let mmpkDirectory: URL
+    /// The currently running download job.
+    @Published private(set) var job: DownloadPreplannedOfflineMapJob?
+    /// The result of the download job.
+    @Published private(set) var result: Result<MobileMapPackage, Error>?
+    
+    init(preplannedMapArea: PreplannedMapArea, offlineMapTask: OfflineMapTask, temporaryDirectory: URL) {
+        self.preplannedMapArea = preplannedMapArea
+        self.offlineMapTask = offlineMapTask
+        self.mmpkDirectory = temporaryDirectory
+            .appendingPathComponent(preplannedMapArea.portalItem.id.rawValue)
+            .appendingPathExtension("mmpk")
+    }
+    
+    deinit {
+        Task { [job] in
+            // Cancel any outstanding job.
+            await job?.cancel()
+        }
```

Comment:
This unstructured task is fine as all it does is cancel the job, which is worth finishing.